icu4c/source/common/wintz.cpp

```diff
@@ -47,77 +48,60 @@ U_NAMESPACE_BEGIN
 U_INTERNAL const char* U_EXPORT2
 uprv_detectWindowsTimeZone()
 {
-    UErrorCode status = U_ZERO_ERROR;
-    char* icuid = nullptr;
-    char dynamicTZKeyName[MAX_TIMEZONE_ID_LENGTH];
-    char tmpid[MAX_TIMEZONE_ID_LENGTH];
-    int32_t len;
-    int id = GEOID_NOT_AVAILABLE;
-    int errorCode;
-    wchar_t ISOcodeW[3] = {}; /* 2 letter ISO code in UTF-16 */
-    char ISOcode[3] = {}; /* 2 letter ISO code in UTF-8 */
-
+    /* Obtain TIME_ZONE_INFORMATION from the API and get the non-localized time zone name. */
     DYNAMIC_TIME_ZONE_INFORMATION dynamicTZI;
     uprv_memset(&dynamicTZI, 0, sizeof(dynamicTZI));
-    uprv_memset(dynamicTZKeyName, 0, sizeof(dynamicTZKeyName));
-    uprv_memset(tmpid, 0, sizeof(tmpid));
 
-    /* Obtain TIME_ZONE_INFORMATION from the API and get the non-localized time zone name. */
-    if (TIME_ZONE_ID_INVALID == GetDynamicTimeZoneInformation(&dynamicTZI)) {
+    DWORD tzIDStatus = GetDynamicTimeZoneInformation(&dynamicTZI);
+    if (tzIDStatus == TIME_ZONE_ID_INVALID || dynamicTZI.TimeZoneKeyName[0] == 0)
         return nullptr;
-    }
 
-    id = GetUserGeoID(GEOCLASS_NATION);
-    errorCode = GetGeoInfoW(id, GEO_ISO2, ISOcodeW, 3, 0);
+    // Is this the correct way to detect cases where the user turns off DST
+    // in Control Panel for a timezone with DST?
+    if (tzIDStatus == TIME_ZONE_ID_UNKNOWN && dynamicTZI.DaylightBias != 0) {
```

Comment:
Since we Vista+ we can call the API `GetDynamicTimeZoneInformation` to populate the struct `DYNAMIC_TIME_ZONE_INFORMATION`. (This is already happening above.)

It turns out that this struct already has a convenient element named `DynamicDaylightTimeDisabled`.
From the documentation:
> this member is TRUE ... if the user has disabled this feature using the Date and Time application in Control Panel.

https://docs.microsoft.com/en-us/windows/win32/api/timezoneapi/ns-timezoneapi-dynamic_time_zone_information

From checking, this struct value should map to the same setting as the registry key 
`DynamicDaylightTimeDisabled` which is under the location `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\TimeZoneInformation`.

So I think we can simply check if `DynamicDaylightTimeDisabled` is non-zero to tell if the user turned off DST or not. :)

Reply:
Thank you, Jeff, for the suggestion. I changed it to use that. 

BTW, I did read the doc on DYNAMIC_TIME_ZONE_INFORMATION and DynamicDaylightTimeDisable field, but it's not clear to me that |DynamicDaylightTimeDisabled=TRUE| corresponds to the user turning off DST manually in the control panel.

Reply:
The cause of my confusion is as following: 

> Indicates whether **dynamic** daylight saving time is disabled. Setting this member to TRUE disables dynamic daylight saving time, causing the system to use **a fixed set of transition dates**.

The above is taken from Windows API doc on the DynamicDaylightTimeDisable field of DYNAMIC_TIME_ZONE_INFORMATION.  It seems to indicate that the field is not about DST being on or off but about whether transition date is **fixed** for all years or the historical change of transition date is taken into account.

Reply:
Hmm, good point @jungshik...
I think I will need to debug and dig-into the actual implementation of the code for the control panel setting in order to verify _for certain_ how it works -- as I think it would be best if the checking was done in the same way. (I hope to have some time tomorrow to look into this more).

Reply:
Okay, after digging into the code (and tracing things in a live Debugger to double-check) I think I now have figured out the exact set of checks that are used to determine if the Control Panel (CPL) thinks that DST is "disabled".

Here is the resulting logic, in a sort of made up pseudo-code to simply things:

```
IF
 (GetDynamicTimeZoneInformation != TIME_ZONE_ID_INVALID && TimeZoneKeyName == Empty && DynamicDaylightTimeDisabled == 1 && StandardDate == DaylightDate && StandardDate != 0)
OR
 (GetDynamicTimeZoneInformation != TIME_ZONE_ID_INVALID && TimeZoneKeyName != NULL  && DynamicDaylightTimeDisabled == 1 && StandardDate == DaylightDate && StandardDate == 0)
THEN
 the DST setting is considered "Disabled".
```

Note that we can factor out the common parts of the logic above in order to remove the redundant checks. (Indeed this is what the code really does). But I wrote it out like this in order to try and be explicit, to make it more clear. (Since I personally find this to be somewhat confusing...).

So I think we could do something like this in the ICU code:

```c++
DYNAMIC_TIME_ZONE_INFORMATION dynamicTZI;
uprv_memset(&dynamicTZI, 0, sizeof(dynamicTZI));
SYSTEMTIME systemTimeZero;
uprv_memset(&systemTimeZero, 0, sizeof(systemTimeZero));

DWORD tzIDStatus = GetDynamicTimeZoneInformation(&dynamicTZI);
if (tzIDStatus == TIME_ZONE_ID_INVALID) {
    return nullptr;
}

// If the DST setting has been turned off in the Control Panel, return "Etc/GMT<offset>".
if (
    (dynamicTZI.DynamicDaylightTimeDisabled != 0) &&
    (uprv_memcmp(&dynamicTZI.StandardDate, &dynamicTZI.DaylightDate, sizeof(dynamicTZI.StandardDate)) == 0) &&
    (
        ((dynamicTZI.TimeZoneKeyName[0] == 0) && (uprv_memcmp(&dynamicTZI.StandardDate, &systemTimeZero, sizeof(systemTimeZero) != 0))) || 
        ((dynamicTZI.TimeZoneKeyName[0] != 0) && (uprv_memcmp(&dynamicTZI.StandardDate, &systemTimeZero, sizeof(systemTimeZero) == 0)))
    )
   )
{
    // DST is disabled.
    //... logic for offset ...
}

// If DST is NOT disabled, but we have an empty TimeZoneKeyName, then it is unclear what we should do, as this shouldn't happen.
if (dynamicTZI.TimeZoneKeyName[0] == 0) {
    return nullptr;
}

```
Note: The placement of the parenthesis above is only to try and help show the logic better. I think for the real code it could be more compact and/or concise. :)

TBH, I'm not really sure if/when/how the TimeZoneKeyName would be *empty*, but nevertheless the code does handle that case, so I think ICU perhaps should too.

I think the more common scenario is this one:
```
TimeZoneKeyName != NULL && DynamicDaylightTimeDisabled == 1 && StandardDate == DaylightDate && StandardDate == 0
```

Looking at the MSDN doc page about the field element `DynamicDaylightTimeDisabled` there is this remark:

> To disable daylight saving time, set this member to TRUE, clear the StandardDate and DaylightDate members, and call SetDynamicTimeZoneInformation.

So the logic in the CPL does seem to match with this.

Reply:
Thank you, Jeff, for figuring this out and sorry for missing your answer to my question. Now that you have PR #1362 , I'm closing this PR.

Reply:
No worries, thanks for creating this PR in the first place! :)